pkg/blockio/blockio.go

```diff
@@ -0,0 +1,58 @@
+// +build linux
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package blockio
+
+import (
+	"github.com/containerd/containerd/log"
+	"github.com/intel/goresctrl/pkg/blockio"
+	"github.com/pkg/errors"
+
+	"github.com/containerd/containerd/plugin"
+)
+
+// Config contains the configuration of the blockio plugin
+type Config struct {
+	ConfigFile string `toml:"config_file" json:"configFile"`
+}
+
+func init() {
+	plugin.Register(&plugin.Registration{
+		Type:   plugin.InternalPlugin,
+		ID:     "blockio",
+		Config: &Config{},
+		InitFn: initBlockio,
+	})
+}
```

Comment:
Rather than adding a new subsystem, can we use NRI? https://github.com/containerd/nri

Reply:
@AkihiroSuda, happy that you asked! Actually I initially planned implementing blockio as an NRI plugin.

However, what changed my mind was the fact that OCI spec [LinuxResources](https://github.com/opencontainers/runtime-spec/blob/master/specs-go/config.go#L366) already had a blockio field, and that runc was already working fine with the CFQ I/O scheduler. Furthermore, [Runc v1.0](https://github.com/opencontainers/runc/releases/tag/v1.0.0)  (released today) supports both CFQ and BFQ schedulers in both cgroups v1 and v2. It obviously provides consistent blockio configuration behavior and low-level error messages to all container runtimes using it.

If NRI allowed a plugin to fill in OCI spec fields and thereby use runc, it would be a clear benefit in architectural perspective. In ideal case NRI would be supported by other container runtimes, too. Then NRI would definitely be the number one option for blockio: there would be a single plugin that you could install to any node running any container runtime that supports NRI. As this is not yet the case, we decided to go with runtime-internal blockio implementations but so that they are very light-weight in the runtime perspective: they use a common class-based blockio configuration and pretty much all the logic is in a library that is common to runtimes.

Related to this, you might be interested in our NRI prototype that enables OCI spec modifications from plugins - among other goodies. But that's a separate discussion.

Reply:
> Rather than adding a new subsystem, can we use NRI? https://github.com/containerd/nri

@AkihiroSuda Sorry if this is slightly off topic, but we have been playing around with a prototype of extending the scope of NRI to allow more versatile plugin functionality, for instance allowing us to implement container resource assignment algorithms as NRI plugins. AFAIK @askervin has taken a look at implementing functionality equivalent to this PR with that extended NRI scope, but we haven't taken a look at whether and how that would be possible with the currently available NRI plugin interface.